File: src/mobile/WhatsTheirTimeMobile/utils/TimeManager.ts
import { City, cities } from '../models/CityData';

// Helper function to get the effective timezone based on current date and DST rules
export const getEffectiveTimezone = (city: City): string => {
  // With IANA timezones, DST is handled automatically, so we can simplify this
  // If there's an alternate timezone defined, we can still use the date range logic
 
    return city.timezone;
 
};

// Helper function to check if a date falls within a range
// Handles cases where the range spans across year boundary (e.g., Oct to Mar)
export const isDateInRange = (
  day: number, 
  month: number, 
  startDay: number, 
  startMonth: number, 
  endDay: number, 
  endMonth: number
): boolean => {
  // Convert dates to simple numeric representation for comparison
  const date = month * 100 + day;
  const start = startMonth * 100 + startDay;
  const end = endMonth * 100 + endDay;
  
  // If range doesn't cross year boundary
  if (start <= end) {
    return date >= start && date <= end;
  } 
  // If range crosses year boundary (e.g., Nov to Mar)
  else {
    return date >= start || date <= end;
  }
};

export const getLocalTime = (city: City | string): string => {
  const timezone = typeof city === 'string' ? city : getEffectiveTimezone(city);
  
  try {
    const date = new Date();
    const formatter = new Intl.DateTimeFormat('en-US', {
      timeZone: timezone,
      hour: 'numeric',
      minute: '2-digit',
      hour12: true
    });
    
    return formatter.format(date);
  } catch (e) {
    // Fallback to local time if timezone is invalid
    return formatTime(new Date());
  }
};

export const getLocalDay = (city: City | string): string => {
  const timezone = typeof city === 'string' ? city : getEffectiveTimezone(city);
  
  try {
    const date = new Date();
    const formatter = new Intl.DateTimeFormat('en-US', {
      timeZone: timezone,
      weekday: 'long'
    });
    
    return formatter.format(date);
  } catch (e) {
    // Fallback to local day
    const days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
    return days[new Date().getDay()];
  }
};

export const getDayDifference = (city: City | string): number => {
  const timezone = typeof city === 'string' ? city : getEffectiveTimezone(city);

  try {
    const now = new Date();
    
    // Get today's date in UTC (year, month, day)
    const utcToday = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate()));

    // Get the date string in the target timezone
    const formatter = new Intl.DateTimeFormat('en-CA', { // Use 'en-CA' for YYYY-MM-DD
      timeZone: timezone,
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
    });
    const targetDateStr = formatter.format(now);
    
    // Parse the target date string into a Date object (in UTC, representing that timezone's date)
    const [year, month, day] = targetDateStr.split('-').map(Number);
    const targetToday = new Date(Date.UTC(year, month - 1, day)); // Month is 0-indexed

    // Calculate the difference in days
    // getTime() returns milliseconds since epoch. Divide by milliseconds in a day.
    const diffMillis = targetToday.getTime() - utcToday.getTime();
    const diffDays = Math.round(diffMillis / (1000 * 60 * 60 * 24));

    return diffDays;
  } catch (e) {
    // console.error("Error in getDayDifference:", e);
    return 0; // Fallback to 0 if there's an error
  }
};

export const getFormattedTimeDifference = (city: City | string): string => {
  const timezone = typeof city === 'string' ? city : getEffectiveTimezone(city);

  try {
    const now = new Date();

    // Get the current time in the target timezone using hour12: false for 24-hour format
    const targetTimeFormatter = new Intl.DateTimeFormat('en-US', {
      timeZone: timezone,
      hour: 'numeric',
      minute: 'numeric',
      hour12: false,
    });
    
    const targetTimeStr = targetTimeFormatter.format(now);
    const [targetHourStr, targetMinuteStr] = targetTimeStr.split(':');
    const targetHour = parseInt(targetHourStr, 10);
    const targetMinute = parseInt(targetMinuteStr, 10);

    // Get the current time in the local timezone (as per system, but mock is UTC)
    const localHour = now.getHours(); // In mock, this is UTC hours (12)
    const localMinute = now.getMinutes(); // In mock, this is UTC minutes (0)

    // Convert both times to minutes from the beginning of the day
    const targetTotalMinutes = targetHour * 60 + targetMinute;
    const localTotalMinutes = localHour * 60 + localMinute;
    
    let diffMinutes = targetTotalMinutes - localTotalMinutes;

    // Get day difference to handle cases where timezone crosses day boundary
    const dayDiff = getDayDifference(timezone);

    // Adjust for day differences
    if (dayDiff === 1) {
      // Target is on the next day
      diffMinutes += 24 * 60;
    } else if (dayDiff === -1) {
      // Target is on the previous day
      diffMinutes -= 24 * 60;
    }

    if (diffMinutes === 0) {
      return 'Same time as you';
    }

    const sign = diffMinutes > 0 ? '+' : '-';
    const absDiffMinutes = Math.abs(diffMinutes);
    const hours = Math.floor(absDiffMinutes / 60);
    const minutes = absDiffMinutes % 60;

    let result = sign;
    if (hours > 0) {
      result += `${hours}h`;
    }
    if (minutes > 0) {
      if (hours > 0) result += ' ';
      result += `${minutes}m`;
    }
    
    return result.trim();

  } catch (e) {
    return 'Unknown';
  }
};

// Helper function to format time as "h:mm AM/PM"
const formatTime = (date: Date): string => {
  let hours = date.getHours();
  const minutes = date.getMinutes();
  const ampm = hours >= 12 ? 'PM' : 'AM';
  
  hours = hours % 12;
  hours = hours ? hours : 12; // the hour '0' should be '12'
  
  const minutesStr = minutes < 10 ? '0' + minutes : minutes;
  
  return `${hours}:${minutesStr} ${ampm}`;
};

File: src/ios/WhatsTheirTime/WhatsTheirTime/Managers/TimeManager.swift
import Foundation

class TimeManager {
    
    static func getLocalTime(for timezone: String) -> String {
        let date = Date()
        let formatter = DateFormatter()
        formatter.dateFormat = "h:mm a"
        
        if let offset = parseGmtOffset(timezone) {
            // Create custom time zone from GMT offset
            let seconds = offset * 60
            let timeZone = TimeZone(secondsFromGMT: seconds)
            formatter.timeZone = timeZone
            return formatter.string(from: date)
        } else {
            // Fallback to system time
            return formatter.string(from: date)
        }
    }
    
    static func getDayDifference(for timezone: String) -> Int {
        let localDate = Date()
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.timeZone = TimeZone.current
        
        let localDateStr = formatter.string(from: localDate)
        
        if let offset = parseGmtOffset(timezone) {
            // Set the formatter to the target timezone
            let seconds = offset * 60
            formatter.timeZone = TimeZone(secondsFromGMT: seconds)
            let remoteDateStr = formatter.string(from: localDate)
            
            // Compare date strings to see if they're different
            if remoteDateStr > localDateStr {
                return 1  // Remote time is a day ahead
            } else if remoteDateStr < localDateStr {
                return -1  // Remote time is a day behind
            }
        }
        
        return 0  // Same day
    }
    
    static func getLocalDay(for timezone: String) -> String {
        let date = Date()
        let formatter = DateFormatter()
        formatter.dateFormat = "EEEE"  // Full day name
        
        if let offset = parseGmtOffset(timezone) {
            // Create custom time zone from GMT offset
            let seconds = offset * 60
            let timeZone = TimeZone(secondsFromGMT: seconds)
            formatter.timeZone = timeZone
            return formatter.string(from: date)
        } else {
            // Fallback to system time
            return formatter.string(from: date)
        }
    }
    
    static func getFormattedTimeDifference(for timezone: String) -> String {
        guard let offset = parseGmtOffset(timezone) else {
            return "Unknown"
        }
        
        // Get local timezone offset in minutes
        let localOffset = TimeZone.current.secondsFromGMT() / 60
        
        // Calculate difference
        let difference = offset - localOffset
        
        if difference == 0 {
            return "Same time as you"
        }
        
        let hours = abs(difference) / 60
        let minutes = abs(difference) % 60
        
        var result = ""
        
        if difference > 0 {
            result = "+"
        } else {
            result = "-"
        }
        
        if hours > 0 {
            result += "\(hours)h"
        }
        
        if minutes > 0 {
            result += " \(minutes)m"
        }
        
        return result
    }
    
    // Parse GMT timezone offset string (e.g., "GMT+5.5") to minutes
    static func parseGmtOffset(_ gmtTimezone: String) -> Int? {
        guard gmtTimezone.hasPrefix("GMT") else {
            return nil
        }
        
        let offsetStr = String(gmtTimezone.dropFirst(3)) // Remove "GMT"
        guard let offsetHours = Double(offsetStr) else {
            return nil
        }
        
        return Int(offsetHours * 60) // Convert to minutes
    }
}